# HG changeset patch
# User Ken Brown <kbrown@cornell.edu>
# Date 1426247865 14400
#      Fri Mar 13 07:57:45 2015 -0400
# Node ID e306f0da362fbef4f307f0ebeceb52c151503528
# Parent  e2f47d3ad784831ace3ab5d4021e5996cf7e1877
Build lisp.dll on Cygwin for use in dynamic loading

* src/spvw.d: Rename "main" to "realmain" on Cygwin for use in
building lisp.dll.  We will then build lisp.exe by compiling
src/lisp.c (which defines "main") and linking it with lisp.dll.
* lisp.c: New file.
* src/makemake.in:
* src/clisp-link.in: Adapt to these changes.

diff -r e2f47d3ad784 -r e306f0da362f src/clisp-link.in
--- a/src/clisp-link.in	Fri Mar 13 07:49:03 2015 -0400
+++ b/src/clisp-link.in	Fri Mar 13 07:57:45 2015 -0400
@@ -83,9 +83,11 @@
 LISPRUN="lisp@LEXE@"
 # ensure that "$1" contains a CLISP linking set
 check_linkset () {
-  test -r "$1"/lisp.a -a -x "$1"/${LISPRUN} -a -r "$1"/lispinit.mem \
-    -a -r "$1"/modules.h -a -r "$1"/modules.o -a -r "$1"/makevars ||
-  fail "directory $1 does not contain a CLISP linking set"
+  local tests="-r $1/lisp.a -a -x $1/${LISPRUN} -a -r $1/lispinit.mem -a -r $1/modules.h -a -r $1/modules.o -a -r $1/makevars"
+  if [ "${HSYSOS}" = cygwin ] ; then
+    tests="${tests} -a -r $1/lisp.c"
+  fi
+  test ${tests} || fail "directory $1 does not contain a CLISP linking set"
 }
 
 # ensure that "$1" contains a CLISP module
@@ -113,7 +115,12 @@
   verbose ${CC} ${CPPFLAGS} ${CFLAGS} -I"$absolute_linkkitdir" -c modules.c
   rm -f modules.c
   # Generate new ${LISPRUN}
-  verbose ${CC} ${CFLAGS} ${CLFLAGS} modules.o ${LIBS} -o ${LISPRUN}
+  if [ "${HSYSOS}" != cygwin ] ; then
+    verbose ${CC} ${CFLAGS} ${CLFLAGS} modules.o ${LIBS} -o ${LISPRUN}
+  else
+    verbose ${CC} -shared ${CFLAGS} ${CLFLAGS} modules.o ${LIBS} -o lisp.dll -Wl,--enable-auto-import -Wl,--out-implib=lisp.dll.a
+    verbose ${CC} ${CFLAGS} ${CLFLAGS} lisp.c lisp.dll.a -o ${LISPRUN}
+  fi
 }
 
 # func_tmpdir
@@ -394,6 +401,9 @@
     if [ -z "$moduledirs" ] ; then
       # Just make links from $destinationdir to $sourcedir
       lncp_some "$sourcedir" "$destinationdir" ${LISPRUN} lispinit.mem modules.h modules.o makevars ${FILES};
+      if [ "${HSYSOS}" = cygwin ] ; then
+        lncp_some "$sourcedir" "$destinationdir" lisp.dll lisp.dll.a
+      fi
     else
       cp "$sourcedir"/modules.h "$destinationdir"/modules.h
       FILES=`lncp_some "$sourcedir" "$destinationdir" ${FILES}`
diff -r e2f47d3ad784 -r e306f0da362f src/lisp.c
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/src/lisp.c	Fri Mar 13 07:57:45 2015 -0400
@@ -0,0 +1,13 @@
+/*
+ * Copyright (C) 2015 Kenneth Brown
+ * This file is a part of GNU CLISP and is distributed under GNU GPL v2+
+ *
+ * This file is compiled only on Cygwin, to build lisp.exe by linking
+ * against lisp.dll.  */
+
+extern int realmain (int argc, char *argv[]); /* Defined in spvw.d.  */
+
+int
+main (int argc, char* argv[]) {
+  return realmain (argc, argv);
+}
diff -r e2f47d3ad784 -r e306f0da362f src/makemake.in
--- a/src/makemake.in	Fri Mar 13 07:49:03 2015 -0400
+++ b/src/makemake.in	Fri Mar 13 07:57:45 2015 -0400
@@ -1545,6 +1545,9 @@
 FILES="${GLLIB_A} "${FILES}
 
 FILES='lisp.a '$FILES
+if [ "${HSYSOS}" = cygwin ] ; then
+  FILES='lisp.c '$FILES
+fi 
 
 CPARTS='        spvw spvwtabf spvwtabs spvwtabo'
 CPARTS=$CPARTS' eval control'
@@ -2663,12 +2666,22 @@
 }
 fi
 
-echol "lisp${LEXE} : ${GLLIB_A} \$(OBJECTS) modules${TOBJ} ${XCL_GETTEXTLIB} data"
+if [ "${HSYSOS}" != cygwin ] ; then
+  echol "lisp${LEXE} : ${GLLIB_A} \$(OBJECTS) modules${TOBJ} ${XCL_GETTEXTLIB} data"
+else
+  echol "lisp.c : ${SRCDIR}lisp.c"
+  echotab "cp ${SRCDIR}lisp.c ."
+  echol
+  echol "lisp${LEXE} lisp.dll lisp.dll.a : ${GLLIB_A} lisp.a modules${TOBJ} ${XCL_GETTEXTLIB} data lisp.c"
+fi
 if [ $HOS != win32 ] ; then
   if [ $XCC_GCC = true -a -n "$SOLARIS_LINKING" ] ; then
     # Dynamically linking on Solaris 2.[23] is a pain.
     LIBGCC_DIR='`'"${XCC} -print-libgcc-file-name"' | sed -e '"'"'s,[^/]*$$,,'"'"'`'
     echotab "${XCC} ${XCFLAGS} ${XCLFLAGS} \$(OBJECTS) modules${TOBJ} ${LIBS} -o lisp${LEXE} || /usr/ccs/bin/ld -V -dy -Bdynamic -Y P,/usr/ccs/lib:/usr/lib -Qy -o lisp${LEXE} ${LIBGCC_DIR}crt1.o ${LIBGCC_DIR}crti.o /usr/ccs/lib/values-Xa.o ${LIBGCC_DIR}crtbegin.o \$(OBJECTS) modules${TOBJ} -L${LIBGCC_DIR} -L/usr/ccs/bin ${LIBS} -lgcc -lc ${LIBGCC_DIR}crtend.o ${LIBGCC_DIR}crtn.o -lgcc"
+  elif [ "${HSYSOS}" = cygwin ] ; then
+    echotab "${XCC} -shared ${XCFLAGS} ${XCLFLAGS} modules${TOBJ} lisp.a ${LIBS} ${OUT}lisp.dll -Wl,--enable-auto-import -Wl,--out-implib=lisp.dll.a"
+    echotab "${XCC} ${XCFLAGS} ${XCLFLAGS} lisp.c lisp.dll.a ${OUT}lisp${LEXE}"
   else
     echotab "${XCC} ${XCFLAGS} ${XCLFLAGS} \$(OBJECTS) modules${TOBJ} ${LIBS} ${OUT}lisp${LEXE}"
   fi
@@ -3159,6 +3172,7 @@
   XCC_CREATESHARED_=`echo "$XCC_CREATESHARED" | sed -e 's,\\$,$$,g'`
   echotab "sed -e 's%@with_dynamic_modules@%${with_dynamic_modules}%' \\"
   echotab "    -e 's%@createsharedlib@%$XCC_CREATESHARED_%' \\"
+  echotab "    -e 's%@HSYSOS@%${HSYSOS}%' \\"
   echotab "    -e 's%@LEXE@%${LEXE}%' -e 's%@SHREXT@%${SHREXT}%' \\"
   if test -z "${CLISP_DEF}"; then
     echotab "    -e 's%@CLISP_DEF@%%' \\"
@@ -3489,7 +3503,7 @@
     echotab "for d in ${line}; do for f in \$\$d/*; do \\"
     echotab "  mkdir -p \$(DESTDIR)\$(lisplibdir)/\$\$d; \\"
     echotab "  case \$\$f in \\"
-    echotab "    */lisp${LEXE}) \$(INSTALL_PROGRAM) \$\$f \$(DESTDIR)\$(lisplibdir)/\$\$f;; \\"
+    echotab "    */lisp${LEXE}|*/lisp${SHREXT}) \$(INSTALL_PROGRAM) \$\$f \$(DESTDIR)\$(lisplibdir)/\$\$f;; \\"
     echotab "    *) \$(INSTALL_DATA) \$\$f \$(DESTDIR)\$(lisplibdir)/\$\$f;; \\"
     echotab "  esac; \\"
     echotab "done; done"
@@ -3880,7 +3894,7 @@
 # without changing the bytecode format and the tables in
 # constobj.d, constpack.d, constsym.d, subr.d, fsubr.d, pseudofun.d.
 clean1 : clean0
-	-\$(RM) lispbibl.h clisp.h *.i *.s *${TOBJ} *.a lisp${LEXE} clisp-link makevars ansi-tests-log ${CLISP_DEF}
+	-\$(RM) lispbibl.h clisp.h *.i *.s *${TOBJ} *.a lisp${LEXE} lisp${SHREXT} lisp.c clisp-link makevars ansi-tests-log ${CLISP_DEF}
 	-\$(RMRF) ${RECOMPILEDIR}
 	-\$(RMRF) ${TESTSDIR}
 	-\$(RMRF) ${ANSITESTSDIR}
diff -r e2f47d3ad784 -r e306f0da362f src/spvw.d
--- a/src/spvw.d	Fri Mar 13 07:49:03 2015 -0400
+++ b/src/spvw.d	Fri Mar 13 07:57:45 2015 -0400
@@ -3664,7 +3664,16 @@
 #ifndef argc_t
   #define argc_t int            /* Type of argc is mostly 'int'. */
 #endif
+
+/* On Cygwin we build lisp.dll with "main" renamed to "realmain".  We
+ will then build lisp.exe by compiling lisp.c (which defines "main")
+ and linking with lisp.dll.  */
+#if defined(UNIX_CYGWIN32)
+  #define main realmain
+#endif
+
 global int main (argc_t argc, char* argv[]) {
+#undef main
   /* initialization of memory management.
    overall procedure:
    process command-line-options.
